torchdata/stateful_dataloader/sampler.py

```diff
@@ -18,58 +19,162 @@ class _StatefulRandomSamplerIterator(Iterator[int], Stateful):
     _GENERATOR = "generator"
     _YIELDED = "yielded"
 
-    def __init__(self, sampler, parent_iterator: Iterator[int]):
+    def __init__(self, sampler):
         self.sampler = sampler
-        self.parent_iterator = parent_iterator
+        self.generator_state = self.sampler.generator.get_state()
         self.yielded = 0
         self.next_yielded = None
-        self.generator_state = sampler.generator.get_state()
+        self.n = len(sampler.data_source)
+        self.replacement = sampler.replacement
+        self.num_samples = sampler.num_samples
+        self.chunk_size = 32
+        self.chunk_index = 0
+        self.perm_index = 0
+        self.perm = None
 
-    def __next__(self) -> int:
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self.replacement:
+            num_full_chunks = self.num_samples // self.chunk_size
+            remainder = self.num_samples % self.chunk_size
+            if self.chunk_index < num_full_chunks:
+                if self.perm is None or not self.perm:
```

Comment:
Why are we doing this chunking logic? I don't think this will give the correct distribution, ie it will depend on num_chunks

Reply:
Hi @andrewkho , 
that's how they are doing it in https://github.com/pytorch/pytorch/blob/b18e3c01aa8bfee37078b4a06cef4361bf63b36b/torch/utils/data/sampler.py#L177
Generate 32 random numbers at once and yield from them. 
In their case `self.num_samples // 32` is essentially num_full_chunks.